# seq-depth

Sequencing depth estimation from fastq files

# about

This is a simple bash script to quickly estimate the sequencing depth for a bunch of fastq files generated by an Illumina sequencer. The script first checks whether you pointed it to fastq files and whether they are gzipped or not. Next, it counts the number of bases in the fastq files (using seqtk if this is installed, otherwise counts are obtained using a bash one-liner). The estimated depth is calculated by dividing the basecounts by the size of the reference genome (user-provided). 

GNU parallel is used to speed up the single-threaded base-counting step.

# example usage

In case you call the script from a directory containing trimmed fastq (or fastq.gz) files:

```Bash
seq_depth_location="/path/to/sequencing_depth.bash"
output_directory="/path/to/output"

bash "$seq_depth_location" \
	-i *trimmed* \
	-g 5000000 \
	-p salmonella \
	-o "$output_directory"
```
# options
- `-i | --input` Path to the input fastq files. If not supplied, the files are expected in the current working directory
- `-g | --genome_size` Size of the reference genome. Default: 500000
- `-p | --prefix` Prefix used to name output files (default 'coolproject')
- `-o | --outdir` Output directory (default '.')
